src/dataflow/src/source/s3.rs

```diff
@@ -752,82 +767,48 @@ async fn download_object(
     };
 
     if let Some(body) = obj.body {
-        let mut reader = body.into_async_read();
-        // unwrap is safe because content length is not allowed to be negative
-        let mut buf = Vec::with_capacity(obj.content_length.unwrap_or(1024).try_into().unwrap());
-
-        match reader.read_to_end(&mut buf).await {
-            Ok(_) => {
-                if buf.is_empty() {
-                    return (DownloadStatus::Ok, None);
-                }
-
-                let bytes_read = buf.len() as u64;
+        // sending EOF after empty objects causes empty records in dataflow
+        let length = if let Some(length) = obj.content_length {
+            if length == 0 {
+                return (DownloadStatus::Empty, None);
+            }
+            length.try_into().unwrap_or(CHUNK_SIZE)
+        } else {
+            log::debug!(
+                "source_id={} surprisingly got not content_length for {}/{}",
+                source_id,
+                bucket,
+                key
+            );
+            CHUNK_SIZE
+        };
```

Comment:
I think it'd be slightly less confusing if we made `read_object_chunk` take an `object_size_if_known: Option<usize>` and did the `unwrap_or` in there.

Reply:
That seems fine.